### PR TITLE
CP-1032: add log patcher to hide PHI and PII information from logs

### PR DESCRIPTION
**Pull request by es-ash**
Estimated review effort: ~30–45 minutes

## Change Summary
Jira ticket: [CP-1032](https://poweredbyash.atlassian.net/browse/CP-1032):

### What it does:
* Adds `PhiPiiLogRedactor`, a Loguru patcher that runs before sinks and redacts PHI/PII from record["message"], structured record["extra"], and exception text wrapped in Loguru’s RecordException.
* Adds `tests/integrations/test_loguru.py` with integration-style tests (real **logger.configure(patcher=...)**) and focused unit tests for edge branches (depth limit, model_dump failure, keyed string redaction, exception fallback, etc.).

## How I know it works

- [x] I reviewed the code myself before requesting another engineer's review

### Testing

* Manual testing: `pytest tests/integrations/test_loguru.py` (substitute any manual Loguru runs you did in an app).
* Automated tests: 27 tests in `tests/integrations/test_loguru.py` — all passed locally (27 passed).
* Coverage spans patcher E2E with logger.add, exception logging, and many branch/unit cases (depth cap, Pydantic, RecordException, string helpers).

## Security

no impact

---

> [!NOTE]
> **Medium Risk**
> Changes logging behavior by mutating Loguru records (message/extra/exception), which could inadvertently over-redact or alter log structures relied on by downstream sinks/alerts. Includes extensive tests, but the surface area (regexes, recursive traversal) makes edge cases possible.
> 
> **Overview**
> Introduces `PhiPiiLogRedactor`, a Loguru `patcher` that redacts PHI/PII from `record["message"]`, structured `record["extra"]` (including nested mappings/collections and Pydantic `model_dump()` objects), and `RecordException` exception messages before logs reach sinks.
> 
> Adds a comprehensive `tests/integrations/test_loguru.py` suite to validate redaction of emails, phones, addresses, auth/secrets/bearer tokens, URL userinfo, and result-payload-like structures, plus fallback behavior on redaction failures and max-depth handling.
> 
> Updates the SonarCloud workflow to use `SonarSource/sonarqube-scan-action@v7` and switches the token env var to `secrets.SONAR_TOKEN`.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 3592a2554961c5d20f66a9f05909adab016b175c. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

[CP-1032]: https://poweredbyash.atlassian.net/browse/CP-1032?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by es-ash** [comment]
bugbot run

**Comment by es-ash** [comment]
bugbot run

**Comment by es-ash** [comment]
bugbot run

**Comment by es-ash** [comment]
bugbot run

**Comment by es-ash** [comment]
bugbot run

**Comment by es-ash** [comment]
bugbot run

File: ash_utils/integrations/loguru.py [hunk 1]
```diff
@@ -0,0 +1,461 @@
...
+    EMAIL_LOCAL_PREFIX_VISIBLE_LEN = 3
+
+    email_pattern: ClassVar[re.Pattern[str]] = re.compile(
+        pattern=r"\b[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}\b",
+        flags=re.IGNORECASE,
+    )
+    url_pattern: ClassVar[re.Pattern[str]] = re.compile(
+        pattern=r"https?://[^\s'\"<>)\]}]+",
+        flags=re.IGNORECASE,
+    )
```
Comment (IceCreamYou) [inline ash_utils/integrations/loguru.py]: This seems problematic, it will prevent us from seeing which endpoint threw an error for example. I don't think we should redact URLs at all, or maybe only the userinfo portion (e.g. `user:password` in `https://user:password@example.com`). If there are query parameters with secrets, they will likely get caught by `secret_value_pattern`.

File: ash_utils/integrations/loguru.py [hunk 2]
```diff
@@ -0,0 +1,461 @@
...
+    sensitive_key_pattern: ClassVar[re.Pattern[str]] = re.compile(
+        pattern=(
+            r"auth|authorization|api_?key|bearer|cookies?|credentials?|pass(?:word|wd)?|private_key|"
+            r"secrets?|sessions?|tokens?"
+        ),
+        flags=re.IGNORECASE,
+    )
+    email_key_pattern: ClassVar[re.Pattern[str]] = re.compile(pattern=r"email", flags=re.IGNORECASE)
```
Comment (IceCreamYou) [inline ash_utils/integrations/loguru.py]: This seems too broad, I would redact keys like `^(?:patient|client|shipping|target|customer|partner|impersonated)?(?:_?user)?_?(?:e?mail|login)(?:_?address)?$` but something like `email_verified` or `has_email` should be fine.

File: tests/integrations/test_loguru.py [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+            "Email john.doe@example.com url https://storage.example.com/result.pdf token=secret-token "
+            "LabResultRow(id=1, value='positive')"
+        )
+
+        output = self.output.getvalue()
+
+        self.assertIn("joh...@example.com", output)
+        self.assertIn("jan...@example.com", output)
+        self.assertIn("KIT123", output)
+        self.assertIn("555-123-4567", output)
```
Comment (IceCreamYou) [inline tests/integrations/test_loguru.py]: Hmm, maybe we should redact values of `^(?:mfa_)?phone(?:_number)?$` keys as well.

File: ash_utils/integrations/loguru.py [hunk 4]
```diff
@@ -0,0 +1,461 @@
...
+        pattern=r"\b[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}\b",
+        flags=re.IGNORECASE,
+    )
+    url_pattern: ClassVar[re.Pattern[str]] = re.compile(
+        pattern=r"https?://[^\s'\"<>)\]}]+",
+        flags=re.IGNORECASE,
+    )
+    secret_value_pattern: ClassVar[re.Pattern[str]] = re.compile(
+        pattern=(
+            r"\b(auth|authorization|api[_-]?key|token|secret|password|passwd|credential|cookie)"
```
Comment (IceCreamYou) [inline ash_utils/integrations/loguru.py]: Maybe also consider adding `|signature|signedheaders` to catch `X-Goog-Signature` and `X-Goog-SignedHeaders` parameters of GCP bucket file URLs.

File: ash_utils/integrations/loguru.py [hunk 5]
```diff
@@ -0,0 +1,461 @@
...
+            r"(?P<quote>['\"])(?P<quoted_key>[A-Za-z_][A-Za-z0-9_]*)(?P=quote)\s*:"
+            r"|(?P<plain_key>[A-Za-z_][A-Za-z0-9_]*)\s*="
+            r")\s*"
+        ),
+        flags=re.IGNORECASE,
+    )
+    sensitive_key_pattern: ClassVar[re.Pattern[str]] = re.compile(
+        pattern=(
+            r"auth|authorization|api_?key|bearer|cookies?|credentials?|pass(?:word|wd)?|private_key|"
+            r"secrets?|sessions?|tokens?"
```
Comment (IceCreamYou) [inline ash_utils/integrations/loguru.py]: We definitely don't want to redact `session_id`/`X-Session-ID`.

I can think of some other possible variations on these keys that we might want to keep, like `token_count` or `password_policy` or `bypass_flag` or `auth_failed_msg`, but I haven't come up with anything else that we actually use.

File: ash_utils/integrations/loguru.py [hunk 6]
```diff
@@ -0,0 +1,461 @@
...
+        flags=re.IGNORECASE,
+    )
+    bearer_token_pattern: ClassVar[re.Pattern[str]] = re.compile(
+        pattern=r"\bbearer\s+[a-z0-9._~+/-]+=*",
+        flags=re.IGNORECASE,
+    )
+    test_result_pattern: ClassVar[re.Pattern[str]] = re.compile(
+        pattern=(
+            r"^(?P<container>(?:desired|expected|target|lab|kit|order|panel|specimen)_(?:result|results)|results?|"
+            r"result_items?|result_entries?|result_lines?|result_records?|result_payload|result_data)$"
```
Comment (IceCreamYou) [inline ash_utils/integrations/loguru.py]: Consider:


```suggestion
            r"result_(?:(?:item|line|record)s?|entry|entries|payload|data))$"
```

Something similar could apply in the next few lines too, combining strings with a `result_` prefix or `_result` suffix.

**Review by IceCreamYou — COMMENTED** [review]
Looks good overall, just a few things that are over-redacted.

File: ash_utils/integrations/loguru.py (hunk 1 above)
Comment (es-ash) [reply ash_utils/integrations/loguru.py]: updated, now only "user:secret" in the URL will be redacted

File: ash_utils/integrations/loguru.py (hunk 2 above)
Comment (es-ash) [reply ash_utils/integrations/loguru.py]: `email_verified` and `has_email` are shown now, added a test case to show that these fields that do not provide actual email info are not redacted

File: tests/integrations/test_loguru.py (hunk 3 above)
Comment (es-ash) [reply tests/integrations/test_loguru.py]: done: `phone_key_pattern` was added + new tests

File: ash_utils/integrations/loguru.py (hunk 4 above)
Comment (es-ash) [reply ash_utils/integrations/loguru.py]: done

File: ash_utils/integrations/loguru.py (hunk 5 above)
Comment (es-ash) [reply ash_utils/integrations/loguru.py]: updated, and new tests provided

File: ash_utils/integrations/loguru.py (hunk 6 above)
Comment (es-ash) [reply ash_utils/integrations/loguru.py]: updated

**Comment by es-ash** [comment]
bugbot run

**Comment by es-ash** [comment]
bugbot run

**Comment by es-ash** [comment]
bugbot run

**Comment by es-ash** [comment]
bugbot run

**Comment by es-ash** [comment]
bugbot run

**Comment by es-ash** [comment]
bugbot run

**Comment by es-ash** [comment]
bugbot run

**Comment by es-ash** [comment]
bugbot run

**Comment by es-ash** [comment]
bugbot run

**Comment by es-ash** [comment]
bugbot run

File: ash_utils/integrations/loguru.py [hunk 7]
```diff
@@ -35,11 +38,11 @@ class PhiPiiLogRedactor:
...
-            r"result_items?|result_entries?|result_lines?|result_records?|result_payload|result_data)$"
-            r"|^(?P<payload>result_code_ids?|result_codes?|result_values?|result_value|observed_values?|"
-            r"numeric_results?|qualitative_results?|interpretations?|reference_ranges?|reference_range|normal_range|"
-            r"abnormal_flags?|units|collection_dates?|collection_date|received_dates?|reported_dates?|date_reported|"
-            r"lab_results?|panel_results?|specimens?|panels?|assays?|analytes?|loinc|order_results?|kit_results?|test_names?)$"
+            r"result_(?:(?:item|line|record)s?|entry|entries|payload|data))$"
+            r"|^(?P<payload>result_code(?:_id)?s?|result_values?|observed_values?|"
+            r"(?:numeric|qualitative|lab|panel|order|kit)_results?|interpretations?|(?:reference|normal)_ranges?|"
+            r"abnormal_flags?|units|(?:collection|received|reported)_dates?|"
+            r"date_reported|specimens?|panels?|assays?|analytes?|loinc|test_names?)$"
```
Comment (IceCreamYou) [inline ash_utils/integrations/loguru.py]: Units and dates should be fine to log.

File: ash_utils/integrations/loguru.py [hunk 8]
```diff
@@ -56,14 +59,37 @@ class PhiPiiLogRedactor:
...
-            r"secrets?|sessions?|tokens?"
+            r"(?:"
+            r"(?:^|_)authorization(?:$|_)"
+            r"|(?:^|_)api_?key(?:$|_)"
+            r"|(?:^|_)bearer(?:$|_)"
+            r"|(?:^|_)cookies?(?:$|_)"
+            r"|(?:^|_)credentials?(?:$|_)"
+            r"|(?:^|_)private_key(?:$|_)"
+            r"|(?:^|_)secrets?(?:$|_)"
+            r"|(?:^|_)pass(?:word|wd)?(?!_policy(?:$|_))(?:$|_)"
```
Comment (IceCreamYou) [inline ash_utils/integrations/loguru.py]: I think the double `(?:$|_)` is redundant

File: ash_utils/integrations/loguru.py [hunk 9]
```diff
@@ -56,14 +59,37 @@ class PhiPiiLogRedactor:
...
+            r"|(?:^|_)sessions?(?!_id(?:$|_))(?:$|_)"
+            r"|(?:^|_)tokens?(?!_count(?:$|_))(?:$|_)"
+            r"|(?:^|_)auth(?!_failed)(?:$|_)"
```
Comment (IceCreamYou) [inline ash_utils/integrations/loguru.py]: It might be better to write this as `(?:^|_)(authorization|api[_-]?key|bearer|...(?:$|_)` instead of repeating the start/end groups for each word.

**Review by IceCreamYou — APPROVED** [review]
Just had some nitpicks, overall LGTM 👍

File: ash_utils/integrations/loguru.py (hunk 7 above)
Comment (es-ash) [reply ash_utils/integrations/loguru.py]: updated

File: ash_utils/integrations/loguru.py (hunk 8 above)
Comment (es-ash) [reply ash_utils/integrations/loguru.py]: updated

File: ash_utils/integrations/loguru.py (hunk 9 above)
Comment (es-ash) [reply ash_utils/integrations/loguru.py]: fixed

**Comment by es-ash** [comment]
bugbot run

**Comment by es-ash** [comment]
bugbot run

**Comment by es-ash** [comment]
bugbot run
